client-v2/src/shared/actions/ArticleFragments.js

```diff
@@ -80,9 +122,32 @@ function addArticleFragment(id: string, supporting: string[] = []) {
       });
 }
 
+const insertAndDedupeSiblings = <T: { id: string, uuid: string }>(
```

Comment:
This is where most of the work is done. I don't like it as it calls dispatch twice which means we get two renders.

The reason for the first dispatch is to run the reducer level logic for insertion, which handles flattening out supporting articles that are too deeply nested as siblings instead of children.

We then call `getState` in the thunk to and take the deduping from there. Potentially we we can bring the other logic into here but it doesn't feel nice as the responsibilities get a bit clouded. Be interest to here @jonathonherbert and @Reettaphant 's opinions on this.

Reply:
The double dispatch is potentially solved by a subscribe debouncer, which we've discussed before. The idea of abstracting the idea of flattening/deduping into their own methods is neat but as you say potentially tricky. I'm not sure it's too big of a deal, but I'm super happy to chat through other ways of getting this done if you'd like to refactor at some point.